File: solutions/java/src/main/java/com/epi/OfflineMinimum.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class OfflineMinimum {
  // @include
  public static List<Integer> offlineMinimum(List<Integer> A, List<Integer> E) {
    List<Integer> R = new ArrayList<>(Collections.nCopies(A.size(), E.size()));
    int pre = 0;

    // Initialize the collection of subsets.
    for (int i = 0; i < E.size(); ++i) {
      for (int j = pre; j <= E.get(i); ++j) {
        R.set(A.get(j), i);
      }
      pre = E.get(i) + 1;
    }

    List<Integer> ret = new ArrayList<>(Collections.nCopies(E.size(), -1));
    List<Integer> set = new ArrayList<>(E.size() + 1); // the disjoint-set
    // initializes the disjoint-set
    for (int i = 0; i < E.size() + 1; ++i) {
      set.add(i);
    }
    for (int i = 0; i < A.size(); ++i) {
      if (findSet(set, R.get(i)) != E.size()
          && ret.get(findSet(set, R.get(i))) == -1) {
        ret.set(set.get(R.get(i)), i);
        unionSet(set, set.get(R.get(i)), set.get(R.get(i)) + 1);
      }
    }
    return ret;
  }

  private static int findSet(List<Integer> set, int x) {
    if (set.get(x) != x) {
      set.set(x, findSet(set, set.get(x))); // path compression.
    }
    return set.get(x);
  }

  private static void unionSet(List<Integer> set, int x, int y) {
    int xRoot = findSet(set, x), yRoot = findSet(set, y);
    set.set(Math.min(xRoot, yRoot), Math.max(xRoot, yRoot));
  }
  // @exclude

  // O(nm) checking method
  static List<Integer> checkAnswer(List<Integer> A, List<Integer> E) {
    boolean[] exist = new boolean[A.size()];
    List<Integer> ans = new ArrayList<>(E.size());

    for (int i = 0; i < E.size(); ++i) {
      int minVal = Integer.MAX_VALUE;
      for (int j = 0; j <= E.get(i); ++j) {
        if (A.get(j) < minVal && !exist[A.get(j)]) {
          minVal = Math.min(A.get(j), minVal);
        }
      }
      exist[minVal] = true;
      ans.add(minVal);
    }
    /*
     * System.out.print("ans2 = "); System.out.println(Arrays.toString(ans));
     */
    return ans;
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n, m;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
        m = gen.nextInt(n) + 1;
      } else if (args.length == 2) {
        n = Integer.parseInt(args[0]);
        m = Integer.parseInt(args[1]);
      } else {
        n = gen.nextInt(1000) + 1;
        m = gen.nextInt(n) + 1;
      }
      System.out.println("n = " + n + ", m = " + m);
      List<Integer> A = new ArrayList<>(n);
      for (int i = 0; i < n; ++i) {
        A.add(i);
      }
      Collections.shuffle(A);

      /*
       * System.out.print("A = "); System.out.println(Arrays.toString(A));
       */
      List<Integer> E = new ArrayList<>(m);
      for (int i = 0; i < m; ++i) {
        E.add(gen.nextInt(n - i) + i);
      }
      Collections.sort(E);
      /*
       * System.out.print("E = "); System.out.println(Arrays.toString(E));
       */
      List<Integer> ans = offlineMinimum(A, E);
      /*
       * System.out.print("ans1 = "); System.out.println(Arrays.toString(ans));
       */
      List<Integer> tmp = checkAnswer(A, E);
      assert(ans.equals(tmp));
    }
  }
}

File: solutions/cpp/Offline_minimum.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <deque>
#include <iostream>
#include <limits>
#include <numeric>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::deque;
using std::endl;
using std::max;
using std::min;
using std::mt19937;
using std::numeric_limits;
using std::ostream_iterator;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

int find_set(int x, vector<int> *set);
void union_set(vector<int> *set, int x, int y);

// @include
vector<int> offline_minimum(const vector<int> &A, const vector<int> &E) {
  vector<int> R(A.size(), E.size());
  int pre = 0;

  // Initialize the collection of subsets.
  for (int i = 0; i < E.size(); ++i) {
    for (int j = pre; j < E[i]; ++j) {
      R[A[j]] = i;
    }
    pre = E[i];
  }

  vector<int> ret(E.size(), -1);  // stores the answer
  vector<int> set(E.size() + 1);  // the disjoint-set
  iota(set.begin(), set.end(), 0);  // initializes the disjoint-set
  for (int i = 0; i < A.size(); ++i) {
    if (find_set(R[i], &set) != E.size() && ret[find_set(R[i], &set)] == -1) {
      ret[set[R[i]]] = i;
      union_set(&set, set[R[i]], set[R[i]] + 1);
    }
  }
  return ret;
}

int find_set(int x, vector<int> *set) {
  if ((*set)[x] != x) {
    (*set)[x] = find_set((*set)[x], set);  // path compression.
  }
  return (*set)[x];
}

void union_set(vector<int> *set, int x, int y) {
  int x_root = find_set(x, set), y_root = find_set(y, set);
  (*set)[min(x_root, y_root)] = max(x_root, y_root);
}
// @exclude

// O(nm) checking method
vector<int> check_answer(const vector<int> &A, const vector<int> &E) {
  deque<bool> exist(A.size(), false);
  vector<int> ans(E.size());

  for (int i = 0; i < E.size(); ++i) {
    int min_val = numeric_limits<int>::max();
    for (int j = 0; j < E[i]; ++j) {
      if (A[j] < min_val && !exist[A[j]]) {
        min_val = min(A[j], min_val);
      }
    }
    exist[min_val] = true;
    ans[i] = min_val;
  }
  /*
  cout << "ans2 = ";
  copy(ans.begin(), ans.end(), ostream_iterator<int>(cout, " "));
  cout << endl;
  */
  return ans;
}

void small_test() {
  vector<int> A = {6, 8, 0, 9, 2, 7, 3, 4, 1, 5};
  vector<int> E = {4, 7, 7, 7, 10};
  auto ans = offline_minimum(A, E);
  copy(ans.begin(), ans.end(), ostream_iterator<int>(cout, " "));
  cout << endl;
  vector<int> expected_ans = {0, 2, 3, 6, 1};
  assert(ans.size() == expected_ans.size() &&
         equal(ans.begin(), ans.end(), expected_ans.begin()));
}

int main(int argc, char *argv[]) {
  small_test();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n, m;
    if (argc == 2) {
      n = atoi(argv[1]);
      uniform_int_distribution<int> dis(1, n);
      m = dis(gen);
    } else if (argc == 3) {
      n = atoi(argv[1]);
      m = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> n_dis(1, 1000);
      n = n_dis(gen);
      uniform_int_distribution<int> m_dis(1, n);
      m = m_dis(gen);
    }
    cout << "n = " << n << ", m = " << m << endl;
    vector<int> A(n);
    iota(A.begin(), A.end(), 0);
    random_device rd;
    mt19937 g(rd());
    shuffle(A.begin(), A.end(), g);
    /*
    cout << "A = ";
    copy(A.begin(), A.end(), ostream_iterator<int>(cout, " "));
    cout << endl;
    */
    vector<int> E;
    for (int i = 0; i < m; ++i) {
      uniform_int_distribution<int> dis(i + 1, n);
      E.emplace_back(dis(gen));
    }
    sort(E.begin(), E.end());
    //*
    cout << "E = ";
    copy(E.begin(), E.end(), ostream_iterator<int>(cout, " "));
    cout << endl;
    //*/
    vector<int> ans = offline_minimum(A, E);
    //*
    cout << "ans1 = ";
    copy(ans.begin(), ans.end(), ostream_iterator<int>(cout, " "));
    cout << endl;
    //*/
    vector<int> tmp = check_answer(A, E);
    assert(ans.size() == tmp.size() &&
           equal(ans.begin(), ans.end(), tmp.begin()));
  }
  return 0;
}
